{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction\n",
    "\n",
    "### Get PhosphPICK Result Files \n",
    "1.  The current version of human proteome (reviewed) sequences were downloaded from [Uniprit.org](https://www.uniprot.org) \n",
    "    - use CreateHumanProteomeDF.ipynb\n",
    "    - Or, go [HERE](https://www.uniprot.org/uniprot/?query=*&fil=organism%3A%22Homo+sapiens+%28Human%29+%5B9606%5D%22+AND+reviewed%3Ayes), click on 'Download'-> FASTA(canonical)\n",
    "2.  split the downloaded fasta file into smaller files with 1000 sequence in each. The last file has 367 sequences. (code in CreateHumanProteomeDF.ipynb)\n",
    "3.  submit the sequence files on [PhosphoPICK](http://bioinf.scmb.uq.edu.au/phosphopick/submit) with the following settings:\n",
    "    - **selection type:** Multiple Kninases\n",
    "    - **species:** human\n",
    "    - **family:** all \n",
    "    - **kinases:** (use the \"shift\" key to select all Kinases in the list)\n",
    "    - **calculate p-value:** checked\n",
    "    - **p-value significance threshold:** no threshold\n",
    "    - **email address:**(enter the email you want to receive the result)\n",
    "4.  download the results from the links in the notification emails, and save them in **'../Data/Raw/PhosphoPICK/'** dir\n",
    "\n",
    "### About the data\n",
    "- Version: Nov., 2019\n",
    "- PhosphoPICK will take about 5 days to complete the prediction and send an email for each of the submitted files with the result downloading link. \n",
    "- the raw predictions are available upon request\n",
    "- PhosphoPICK result files contain the following columns:\n",
    "    - **identifier:** identifiers from the fasta files submitted for PhosphoPICK prediction\t\n",
    "    - **Uniprot-Acc:** uniprotID of the protein that has been found by the BLAST search (please note if multiple proteins return as the most significant match PhosphoPICK will return all matching Uniprot IDs).\n",
    "    - **Uniprot-Acc:** uniprotID of the protein that has been found by the BLAST search (please note if multiple proteins return as the most significant match PhosphoPICK will return all matching Uniprot IDs).\n",
    "    - **blastp-identity:** blastp-identity of the predicted substrate to the submitted seq\n",
    "    - **kinase:** the kinase being scored for this substrate/site.\n",
    "    - **context-score:** probability according to the PhosphoPICK Bayesian network model that your chosen kinase is phosphorylating the protein.\n",
    "    - **context-p-value**\t\n",
    "    - **site:** the location of a potential phosphorylation site on the protein.\n",
    "    - **site-score:** probability according to the naive Bayes model that this site is being phosphorylated by the kinase.\t\n",
    "    - **site-p-value**\t\n",
    "    - **combined-score:** Represents the combined probability according to the context model and the sequence model that this site is phosphorylated by the kinase. Calculated as the average of the context score and the site score.\n",
    "    - **combined-p-value**\n",
    "\n",
    "\n",
    "        \n",
    "### Preprocessing PhosphoPICK Raw\n",
    "In order to convert this to standard format, here we:\n",
    "1. Filtering Date: \n",
    "    - remove isoforms\n",
    "    - remove predicted substrate with blastp-identity != 100.0 \n",
    "        - only keep predictions that the predicted substrate is the input protein submitted for PhosphoPICK prediction\n",
    "2. Mapping accessions:\n",
    "    - **Mapping substrate accessions:**  \n",
    "        - extrated the uniprotID from 'identifier' column (this is the uniprotID of the protein submitted for PhosphoPICK prediction)\n",
    "        - using the 'Uniprot-Acc' (the uniprotID of the predicted substrate by PhosphoPICK) column to retrieve the current uniprotID for the predicted substrate\n",
    "        - remove data that the uniprotID of the protein submitted for PhosphoPICK prediction is different from the retrieved current uniprotID for the predicted substrate (this will remove any data with outdated uniprotID, and perdiction with substrate has seq with 100% blastp-identity but is not the input portein)\n",
    "    - **Mapping kinase accessions:** \n",
    "        - get uniprotID for the kinases using 'kinase' column \n",
    "3. **Mapping sites:**\n",
    "    - get the +/- 7 AA peptide seq from the reference human proteome\n",
    "    - catch outdated records causing by: deleted uniprotID, updated UniprotID, sequence of the given uniprotID changed causing out of range error...\n",
    "    \n",
    "\n",
    "### Creating Resource Files\n",
    "1.  **'globalKinaseMap.csv':** After step 3 in preprocessing, add unique kinases from PhosphoPICK to the globel kinase resource. \n",
    "    \n",
    "\n",
    "### Standard Formatted PhosphoPICK\n",
    "**'FormattedPhosphoPICK.csv':** Standardize the preprocessed file with following columns:\n",
    "- **substrate_id** - unique IDs for the substrate phosphorylation site (substrate_acc + position)\n",
    "- **substrate** - gene name for the substrates\n",
    "- **substrate_acc** - mapped UniprotIDs for the substrates\n",
    "- **site** - phosphorylation  site\n",
    "- **pep** - +/- 7 AA peptide sequence around the site\n",
    "- **kinase** - Kinase name"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Initializing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# IMPORTS\n",
    "import pandas as pd\n",
    "import os\n",
    "import re\n",
    "import glob\n",
    "\n",
    "import humanProteomesReference, phosphoPick_convert, getUniprotID, checkSite\n",
    "\n",
    "#only need when testing the code\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# DEFINE FILE NAMES/DIRs\n",
    "##################\n",
    "# Version (Date) #\n",
    "##################\n",
    "version = '2019-12-11'\n",
    "\n",
    "##################\n",
    "# File Location  #\n",
    "##################\n",
    "# local (../../)\n",
    "base = '../../'\n",
    "\n",
    "####################################################\n",
    "# For Prepare Fasta Files to Submit in PhosphoPICK #\n",
    "####################################################\n",
    "\n",
    "# Human Proteome fasta file\n",
    "HP_fasta = base + 'Data/Raw/HumanProteome/humanProteome_' + version + '.fasta'\n",
    "# Dir for splited Human Proteome fasta files\n",
    "HP_dir = base + 'Data/Raw/HumanProteome/'\n",
    "\n",
    "\n",
    "# human proteome referece file \n",
    "HP_csv = base + 'Data/Map/humanProteome_' + version + '.csv'\n",
    "\n",
    "####################################################\n",
    "# For Preprocessing PhosphoPICK Prediction Results #\n",
    "# (for the Nov 2019 results)                       #\n",
    "#--------------------------------------------------#\n",
    "# . The files submitted for PhosphoPICK predictor  #\n",
    "#   is NOT the up-to-date human proteom sequences  #\n",
    "# . There has been an update in human proteomes    #\n",
    "#   from the time the perdiction results were got  #\n",
    "#   to running the preporcessing steps.            #\n",
    "####################################################\n",
    "\n",
    "# PhosphoPICK results dir\n",
    "PP_dir = base + 'Data/Raw/PhosphoPICK/'\n",
    "PP_update_dir = base + 'Data/Raw/PhosphoPICK/updated/'\n",
    "\n",
    "# PhosphoPICK temp dir\n",
    "PP_temp_dir_acc = base + 'Data/Temp/PhosphoPICK/mappedAcc/'\n",
    "PP_temp_dir_site = base + 'Data/Temp/PhosphoPICK/mappedSite/'\n",
    "PP_temp_dir_acc_update = base + 'Data/Temp/PhosphoPICK/mappedAcc/updated/'\n",
    "PP_temp_dir_site_update = base + 'Data/Temp/PhosphoPICK/mappedSite/updated/'\n",
    "\n",
    "# Resource Files\n",
    "HK_org = base + 'Data/Raw/HumanKinase/globalKinaseMap.txt'                  # orginal manually created kinase file\n",
    "KinaseMap = base + 'Data/Map/globalKinaseMap.csv'                           # add all unique kinase in HPRD to the global file\n",
    "\n",
    "# Standard formatted output file\n",
    "PP_formatted = base + 'Data/Formatted/PhosphoPICK/PhosphoPICK_formatted_' + version + '.csv'       # preprocessed file with cloumns: substrate_id/substrate/substrate_acc/kinase/site/pep/score\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preprocessing PhosphoPICK Raw"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mapping Accessions (UniprotID) and Site\n",
    "1. Filtering Date: \n",
    "    - remove isoforms\n",
    "    - remove predicted substrate with blastp-identity != 100.0 \n",
    "        - only keep predictions that the predicted substrate is the input protein submitted for PhosphoPICK prediction\n",
    "2. Mapping accessions:\n",
    "    - **Mapping substrate accessions:**  \n",
    "        - extrated the uniprotID from 'identifier' column (this is the uniprotID of the protein submitted for PhosphoPICK prediction)\n",
    "        - using the 'Uniprot-Acc' (the uniprotID of the predicted substrate by PhosphoPICK) column to retrieve the current uniprotID for the predicted substrate\n",
    "        - remove data that the uniprotID of the protein submitted for PhosphoPICK prediction is different from the retrieved current uniprotID for the predicted substrate (this will remove any data with outdated uniprotID, and perdiction with substrate has seq with 100% blastp-identity but is not the input portein)\n",
    "    - **Mapping kinase accessions:** \n",
    "        - get uniprotID for the kinases using 'kinase' column \n",
    "3. **Mapping sites:**\n",
    "    - get the +/- 7 AA peptide seq from the reference human proteome\n",
    "    - catch outdated records causing by: deleted uniprotID, updated UniprotID, sequence of the given uniprotID changed causing out of range error...\n",
    "\n",
    "**Output Files Dataframe:**\n",
    "- **Uniprot-Acc:** the uniprotID of the predicted substrate by PhosphoPICK\n",
    "- **substrate_id:** unique IDs for the substrate phosphorylation site (substrate_acc + position)\n",
    "- **substrate_acc:** substrate uniprotID\n",
    "- **site:** aa + position in protein sequence\n",
    "- **position:** position in protein sequence\n",
    "- **pep:** +/- 7 AA\n",
    "- **kinase:** kinase name used in PhosphoPICK\n",
    "- **kinase_acc:** kinase uniprotID\n",
    "- **combined-p-value:** PhosphoPICK score"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Mapping Accessions**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "reading  13001-14000.txt\n",
      "getting unique sub\n",
      "getting sub_acc\n",
      "merge\n",
      "done 458.9465310573578\n",
      "getting unique kin\n",
      "getting kin_acc\n",
      "RPSK6A5 no hit in human\n",
      "MAP3KB no hit in human\n",
      "merge\n",
      "done 45.92072534561157\n",
      "saving\n",
      "Done 25.648595094680786\n",
      "reading  3001-4000.txt\n",
      "getting unique sub\n",
      "getting sub_acc\n",
      "merge\n",
      "done 442.121808052063\n",
      "getting unique kin\n",
      "getting kin_acc\n",
      "RPSK6A5 no hit in human\n",
      "MAP3KB no hit in human\n",
      "merge\n",
      "done 47.999385833740234\n",
      "saving\n",
      "Done 22.979498863220215\n",
      "reading  1-1000.txt\n",
      "getting unique sub\n",
      "getting sub_acc\n",
      "Q5JT78 no hit in human\n",
      "merge\n",
      "done 399.5749170780182\n",
      "getting unique kin\n",
      "getting kin_acc\n",
      "RPSK6A5 no hit in human\n",
      "MAP3KB no hit in human\n",
      "merge\n",
      "done 44.98341202735901\n",
      "saving\n",
      "Done 21.85321807861328\n",
      "reading  15001-16000.txt\n",
      "getting unique sub\n",
      "getting sub_acc\n",
      "P0CZ20 no hit in human\n",
      "B3EWG4 no hit in human\n",
      "B3EWG4 no hit in human\n",
      "merge\n",
      "done 388.16595697402954\n",
      "getting unique kin\n",
      "getting kin_acc\n",
      "RPSK6A5 no hit in human\n",
      "MAP3KB no hit in human\n",
      "merge\n",
      "done 41.36503720283508\n",
      "saving\n",
      "Done 21.812942028045654\n",
      "reading  10001-11000.txt\n",
      "getting unique sub\n",
      "getting sub_acc\n",
      "Q5JXX5 no hit in human\n",
      "merge\n",
      "done 340.52795600891113\n",
      "getting unique kin\n",
      "getting kin_acc\n",
      "RPSK6A5 no hit in human\n",
      "MAP3KB no hit in human\n",
      "merge\n",
      "done 33.0019588470459\n",
      "saving\n",
      "Done 23.31880211830139\n",
      "reading  16001-17000.txt\n",
      "getting unique sub\n",
      "getting sub_acc\n",
      "B3EWG4 no hit in human\n",
      "merge\n",
      "done 329.76578402519226\n",
      "getting unique kin\n",
      "getting kin_acc\n",
      "RPSK6A5 no hit in human\n",
      "MAP3KB no hit in human\n",
      "merge\n",
      "done 43.6525502204895\n",
      "saving\n",
      "Done 24.139140844345093\n",
      "reading  8001-9000.txt\n",
      "getting unique sub\n",
      "getting sub_acc\n",
      "merge\n",
      "done 468.32062888145447\n",
      "getting unique kin\n",
      "getting kin_acc\n",
      "RPSK6A5 no hit in human\n",
      "MAP3KB no hit in human\n",
      "merge\n",
      "done 56.35699677467346\n",
      "saving\n",
      "Done 28.59500813484192\n",
      "reading  2001-3000.txt\n",
      "getting unique sub\n",
      "getting sub_acc\n",
      "merge\n",
      "done 440.55045318603516\n",
      "getting unique kin\n",
      "getting kin_acc\n",
      "RPSK6A5 no hit in human\n",
      "MAP3KB no hit in human\n",
      "merge\n",
      "done 48.410074949264526\n",
      "saving\n",
      "Done 27.257181882858276\n",
      "reading  5001-6000.txt\n",
      "getting unique sub\n",
      "getting sub_acc\n",
      "merge\n",
      "done 411.73199796676636\n",
      "getting unique kin\n",
      "getting kin_acc\n",
      "RPSK6A5 no hit in human\n",
      "MAP3KB no hit in human\n",
      "merge\n",
      "done 48.64367389678955\n",
      "saving\n",
      "Done 25.9805269241333\n",
      "reading  4001-5000.txt\n",
      "getting unique sub\n",
      "getting sub_acc\n",
      "merge\n",
      "done 449.4707248210907\n",
      "getting unique kin\n",
      "getting kin_acc\n",
      "RPSK6A5 no hit in human\n",
      "MAP3KB no hit in human\n",
      "merge\n",
      "done 53.620739698410034\n",
      "saving\n",
      "Done 28.956773042678833\n",
      "reading  6001-7000.txt\n",
      "getting unique sub\n",
      "getting sub_acc\n",
      "merge\n",
      "done 494.44567012786865\n",
      "getting unique kin\n",
      "getting kin_acc\n",
      "RPSK6A5 no hit in human\n",
      "MAP3KB no hit in human\n",
      "merge\n",
      "done 50.80067014694214\n",
      "saving\n",
      "Done 27.15272307395935\n",
      "reading  1001-2000.txt\n",
      "getting unique sub\n",
      "getting sub_acc\n",
      "merge\n",
      "done 438.56801986694336\n",
      "getting unique kin\n",
      "getting kin_acc\n",
      "RPSK6A5 no hit in human\n",
      "MAP3KB no hit in human\n",
      "merge\n",
      "done 50.79615521430969\n",
      "saving\n",
      "Done 27.587910890579224\n",
      "reading  7001-8000.txt\n",
      "getting unique sub\n",
      "getting sub_acc\n",
      "merge\n",
      "done 468.66574716567993\n",
      "getting unique kin\n",
      "getting kin_acc\n",
      "RPSK6A5 no hit in human\n",
      "MAP3KB no hit in human\n",
      "merge\n",
      "done 54.072288036346436\n",
      "saving\n",
      "Done 27.589793920516968\n",
      "reading  14001-15000.txt\n",
      "getting unique sub\n",
      "getting sub_acc\n",
      "merge\n",
      "done 508.6065797805786\n",
      "getting unique kin\n",
      "getting kin_acc\n",
      "RPSK6A5 no hit in human\n",
      "MAP3KB no hit in human\n",
      "merge\n",
      "done 56.11255884170532\n",
      "saving\n",
      "Done 23.87323498725891\n",
      "reading  19001-20350.txt\n",
      "getting unique sub\n",
      "getting sub_acc\n",
      "Q9UGB4 no hit in human\n",
      "A6NEA5 no hit in human\n",
      "merge\n",
      "done 644.0101702213287\n",
      "getting unique kin\n",
      "getting kin_acc\n",
      "RPSK6A5 no hit in human\n",
      "MAP3KB no hit in human\n",
      "merge\n",
      "done 53.497946977615356\n",
      "saving\n",
      "Done 31.776172876358032\n",
      "reading  11001-12000.txt\n",
      "getting unique sub\n",
      "getting sub_acc\n",
      "merge\n",
      "done 467.24060583114624\n",
      "getting unique kin\n",
      "getting kin_acc\n",
      "RPSK6A5 no hit in human\n",
      "MAP3KB no hit in human\n",
      "merge\n",
      "done 52.390093088150024\n",
      "saving\n",
      "Done 24.29642081260681\n",
      "reading  17001-18000.txt\n",
      "getting unique sub\n",
      "getting sub_acc\n",
      "Q6ZN80 no hit in human\n",
      "A8MUQ0 no hit in human\n",
      "merge\n",
      "done 457.16089701652527\n",
      "getting unique kin\n",
      "getting kin_acc\n",
      "RPSK6A5 no hit in human\n",
      "MAP3KB no hit in human\n",
      "merge\n",
      "done 52.4887011051178\n",
      "saving\n",
      "Done 23.839888095855713\n",
      "reading  18001-19000.txt\n",
      "getting unique sub\n",
      "getting sub_acc\n",
      "merge\n",
      "done 439.6862518787384\n",
      "getting unique kin\n",
      "getting kin_acc\n",
      "RPSK6A5 no hit in human\n",
      "MAP3KB no hit in human\n",
      "merge\n",
      "done 47.016953229904175\n",
      "saving\n",
      "Done 22.01080012321472\n",
      "reading  9001-10000.txt\n",
      "getting unique sub\n",
      "getting sub_acc\n",
      "merge\n",
      "done 423.4310703277588\n",
      "getting unique kin\n",
      "getting kin_acc\n",
      "RPSK6A5 no hit in human\n",
      "MAP3KB no hit in human\n",
      "merge\n",
      "done 50.95588278770447\n",
      "saving\n",
      "Done 23.666645050048828\n",
      "reading  12001-13000.txt\n",
      "getting unique sub\n",
      "getting sub_acc\n",
      "merge\n",
      "done 457.99590015411377\n",
      "getting unique kin\n",
      "getting kin_acc\n",
      "RPSK6A5 no hit in human\n",
      "MAP3KB no hit in human\n",
      "merge\n",
      "done 51.28110909461975\n",
      "saving\n",
      "Done 24.361382007598877\n",
      "Time\t10664.738\n"
     ]
    }
   ],
   "source": [
    "convert_type = 'acc'\n",
    "phosphoPick_convert.pick_convert_directory(PP_dir, 'na', PP_temp_dir_acc, convert_type)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Mapping Site**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Set input file dir...\n",
      "done\n",
      "read the Human Proteome df...\n",
      "done\n",
      "processing  13001-14000 .txt\n",
      "Saving  13001-14000\n",
      "13001-14000 155.31652522087097\n",
      "processing  3001-4000 .txt\n",
      "Saving  3001-4000\n",
      "3001-4000 135.85251212120056\n",
      "processing  1-1000 .txt\n",
      "Saving  1-1000\n",
      "1-1000 128.94466519355774\n",
      "processing  15001-16000 .txt\n",
      "Saving  15001-16000\n",
      "15001-16000 135.2762269973755\n",
      "processing  10001-11000 .txt\n",
      "Saving  10001-11000\n",
      "10001-11000 133.8019778728485\n",
      "processing  16001-17000 .txt\n",
      "Saving  16001-17000\n",
      "16001-17000 137.52903985977173\n",
      "processing  8001-9000 .txt\n",
      "Saving  8001-9000\n",
      "8001-9000 136.03232216835022\n",
      "processing  2001-3000 .txt\n",
      "Saving  2001-3000\n",
      "2001-3000 131.88873171806335\n",
      "processing  5001-6000 .txt\n",
      "Saving  5001-6000\n",
      "5001-6000 149.10758900642395\n",
      "processing  4001-5000 .txt\n",
      "Saving  4001-5000\n",
      "4001-5000 152.20608711242676\n",
      "processing  6001-7000 .txt\n",
      "Saving  6001-7000\n",
      "6001-7000 149.08443593978882\n",
      "processing  1001-2000 .txt\n",
      "Saving  1001-2000\n",
      "1001-2000 149.21809601783752\n",
      "processing  7001-8000 .txt\n",
      "Saving  7001-8000\n",
      "7001-8000 137.2006869316101\n",
      "processing  14001-15000 .txt\n",
      "Saving  14001-15000\n",
      "14001-15000 142.13541984558105\n",
      "processing  19001-20350 .txt\n",
      "Saving  19001-20350\n",
      "19001-20350 182.6816849708557\n",
      "processing  11001-12000 .txt\n",
      "Saving  11001-12000\n",
      "11001-12000 129.362863779068\n",
      "processing  17001-18000 .txt\n",
      "Saving  17001-18000\n",
      "17001-18000 141.67677116394043\n",
      "processing  18001-19000 .txt\n",
      "Saving  18001-19000\n",
      "18001-19000 134.71866011619568\n",
      "processing  9001-10000 .txt\n",
      "Saving  9001-10000\n",
      "9001-10000 138.90637683868408\n",
      "processing  12001-13000 .txt\n",
      "Saving  12001-13000\n",
      "12001-13000 147.50381708145142\n",
      "Time\t2848.783\n"
     ]
    }
   ],
   "source": [
    "convert_type = 'site'\n",
    "phosphoPick_convert.pick_convert_directory(PP_temp_dir_acc, HP_csv, PP_temp_dir_site, convert_type)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Remove unmapped/outdated results**\n",
    "- Check if there is any unmapped substrate/site due to outdated uniprot sequence records\n",
    "    - get a list of outdated uniprotID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_results = glob.glob(PP_temp_dir_site + '*.csv')\n",
    "    \n",
    "updatedSub_li = []\n",
    "for filename in all_results:\n",
    "    # uncomment the next line if df_subMap was not defined\n",
    "    df_unmapped = pd.read_csv(filename, usecols = ['substrate_id','substrate_acc', 'site'])\n",
    "    df_unmapped = df_unmapped[~(df_unmapped['site'].str.contains('S|T|Y', na=False)) | (df_unmapped['substrate_id'] == 'outdated')]\n",
    "    df_unmapped = df_unmapped.substrate_acc.drop_duplicates()\n",
    "    unmapped_li = df_unmapped.values.tolist()\n",
    "    updatedSub_li.extend(unmapped_li)\n",
    "\n",
    "updatedSub_li = list(dict.fromkeys(updatedSub_li))\n",
    "updatedSub_li "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Q6QHF9',\n",
       " 'Q9H4K1',\n",
       " 'O43151',\n",
       " 'Q9C0I4',\n",
       " 'A8MYV0',\n",
       " 'Q9HBT7',\n",
       " 'Q06124',\n",
       " 'P08913',\n",
       " 'Q8IXR5']"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_results = glob.glob(PP_temp_dir_site + '*.csv')\n",
    "    \n",
    "updatedSub_li = []\n",
    "for filename in all_results:\n",
    "    # uncomment the next line if df_subMap was not defined\n",
    "    df_unmapped = pd.read_csv(filename, usecols = ['substrate_id','substrate_acc', 'site'])\n",
    "    df_unmapped = df_unmapped[~(df_unmapped['site'].str.contains('S|T|Y', na=False)) | (df_unmapped['substrate_id'] == 'outdated')]\n",
    "    df_unmapped = df_unmapped.substrate_acc.drop_duplicates()\n",
    "    unmapped_li = df_unmapped.values.tolist()\n",
    "    updatedSub_li.extend(unmapped_li)\n",
    "\n",
    "updatedSub_li = list(dict.fromkeys(updatedSub_li))\n",
    "updatedSub_li "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Remove any record with unmapped substrate_acc/site in '*_mappedSite.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chunk time\t35.869\n",
      "chunk time\t37.469\n",
      "chunk time\t36.316\n",
      "chunk time\t35.933\n",
      "chunk time\t34.634\n",
      "chunk time\t34.238\n",
      "chunk time\t36.334\n",
      "chunk time\t34.589\n",
      "chunk time\t36.695\n",
      "chunk time\t37.314\n",
      "chunk time\t40.518\n",
      "chunk time\t35.472\n",
      "chunk time\t37.720\n",
      "chunk time\t33.727\n",
      "chunk time\t34.665\n",
      "chunk time\t33.893\n",
      "chunk time\t33.305\n",
      "chunk time\t44.794\n",
      "chunk time\t32.537\n",
      "chunk time\t34.160\n"
     ]
    }
   ],
   "source": [
    "all_results = glob.glob(PP_temp_dir_site + '*.csv')\n",
    "for filename in all_results:\n",
    "    start = time.time()\n",
    "    df_mapSite = pd.read_csv(filename)\n",
    "    # remove the outdated records from df_subMap\n",
    "    df_update = df_mapSite[~df_mapSite['substrate_acc'].isin(updatedSub_li)]\n",
    "    df_update.to_csv(filename, chunksize=100000, index=False)\n",
    "    end = time.time()\n",
    "    print (f\"chunk time\\t{(end-start):.3f}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Update PhosphoPICK results\n",
    "- The next 6 cells is only for PhosphoPICK results from outdated Human Proteomes sequences (input sequences for PhosphoPICK prediction an earlier version than the referece human proteome sequence)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. download the sequence fasta file of the above UniprotID (updatedSub_li) from Uniprot.org, save it as '../Data/Raw/HumanProteome/phosphoPICK_updateSub.fasta'. Submit the 'phosphoPICK_updateSub.fasta' on PhosphoPICK again.\n",
    "2. run mapAcc and mapSite function for the result file from 'phosphoPICK_updateSub.fasta'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "pp_update_hp = base + 'Data/Raw/HumanProteome/phosphoPICK_updateSub.fasta'\n",
    "update_hp_csv = base + 'Data/Map/phosphoPICK_updateSub.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "reading  PICK_updated-9.txt\n",
      "getting unique sub\n",
      "getting sub_acc\n",
      "merge\n",
      "done 2.027851104736328\n",
      "getting unique kin\n",
      "getting kin_acc\n",
      "RPSK6A5 no hit in human\n",
      "MAP3KB no hit in human\n",
      "merge\n",
      "done 34.19140601158142\n",
      "saving\n",
      "Done 0.16403412818908691\n"
     ]
    }
   ],
   "source": [
    "convert_type = 'acc'\n",
    "phosphoPick_convert.pick_convert_directory(PP_update_dir, 'na', PP_temp_dir_acc_update, convert_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Formatting  PICK_updated-9 ...\n",
      "Reading input file...\n",
      "Get unique substrate sites...\n",
      "Done. Time\t21.207\n"
     ]
    }
   ],
   "source": [
    "humanProteomesReference.fastaToCSV(pp_update_hp, update_hp_csv)\n",
    "\n",
    "convert_type = 'site'\n",
    "phosphoPick_convert.pick_convert_directory(PP_temp_dir_acc_update, update_hp_csv, PP_temp_dir_site_update, convert_type)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3. save 'PICK_updated-9_mappedSite.csv' under the same dir as other *_mappedSite.csv files\n",
    "    - remove any unmapped sites, if any"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uniprot-Acc</th>\n",
       "      <th>blastp-identity</th>\n",
       "      <th>kinase</th>\n",
       "      <th>position</th>\n",
       "      <th>combined-p-value</th>\n",
       "      <th>substrate_acc</th>\n",
       "      <th>kinase_acc</th>\n",
       "      <th>site</th>\n",
       "      <th>pep</th>\n",
       "      <th>substrate_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>O43151</td>\n",
       "      <td>100.0</td>\n",
       "      <td>CDK1</td>\n",
       "      <td>396</td>\n",
       "      <td>0.001721</td>\n",
       "      <td>O43151</td>\n",
       "      <td>P06493</td>\n",
       "      <td>S396</td>\n",
       "      <td>SPPAPFRSPQSYLRA</td>\n",
       "      <td>O43151_396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>O43151</td>\n",
       "      <td>100.0</td>\n",
       "      <td>CDK1</td>\n",
       "      <td>1549</td>\n",
       "      <td>0.003608</td>\n",
       "      <td>O43151</td>\n",
       "      <td>P06493</td>\n",
       "      <td>S1549</td>\n",
       "      <td>FNSALKGSPGFQDKL</td>\n",
       "      <td>O43151_1549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>O43151</td>\n",
       "      <td>100.0</td>\n",
       "      <td>CDK1</td>\n",
       "      <td>753</td>\n",
       "      <td>0.006664</td>\n",
       "      <td>O43151</td>\n",
       "      <td>P06493</td>\n",
       "      <td>S753</td>\n",
       "      <td>ESPFATRSPKQIKIE</td>\n",
       "      <td>O43151_753</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>O43151</td>\n",
       "      <td>100.0</td>\n",
       "      <td>CDK1</td>\n",
       "      <td>1000</td>\n",
       "      <td>0.007765</td>\n",
       "      <td>O43151</td>\n",
       "      <td>P06493</td>\n",
       "      <td>T1000</td>\n",
       "      <td>CKYARSKTPRKFRLA</td>\n",
       "      <td>O43151_1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>O43151</td>\n",
       "      <td>100.0</td>\n",
       "      <td>CDK1</td>\n",
       "      <td>428</td>\n",
       "      <td>0.008986</td>\n",
       "      <td>O43151</td>\n",
       "      <td>P06493</td>\n",
       "      <td>T428</td>\n",
       "      <td>SSAFPPATPRTEFPE</td>\n",
       "      <td>O43151_428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42365</th>\n",
       "      <td>P08913</td>\n",
       "      <td>100.0</td>\n",
       "      <td>PRKDC</td>\n",
       "      <td>242</td>\n",
       "      <td>0.798014</td>\n",
       "      <td>P08913</td>\n",
       "      <td>P78527</td>\n",
       "      <td>T242</td>\n",
       "      <td>YQIAKRRTRVPPSRR</td>\n",
       "      <td>P08913_242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42366</th>\n",
       "      <td>P08913</td>\n",
       "      <td>100.0</td>\n",
       "      <td>PRKDC</td>\n",
       "      <td>247</td>\n",
       "      <td>0.808342</td>\n",
       "      <td>P08913</td>\n",
       "      <td>P78527</td>\n",
       "      <td>S247</td>\n",
       "      <td>RRTRVPPSRRGPDAV</td>\n",
       "      <td>P08913_247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42367</th>\n",
       "      <td>P08913</td>\n",
       "      <td>100.0</td>\n",
       "      <td>PRKDC</td>\n",
       "      <td>135</td>\n",
       "      <td>0.817593</td>\n",
       "      <td>P08913</td>\n",
       "      <td>P78527</td>\n",
       "      <td>S135</td>\n",
       "      <td>DVLFCTSSIVHLCAI</td>\n",
       "      <td>P08913_135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42368</th>\n",
       "      <td>P08913</td>\n",
       "      <td>100.0</td>\n",
       "      <td>PRKDC</td>\n",
       "      <td>143</td>\n",
       "      <td>0.826284</td>\n",
       "      <td>P08913</td>\n",
       "      <td>P78527</td>\n",
       "      <td>S143</td>\n",
       "      <td>IVHLCAISLDRYWSI</td>\n",
       "      <td>P08913_143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42369</th>\n",
       "      <td>P08913</td>\n",
       "      <td>100.0</td>\n",
       "      <td>PRKDC</td>\n",
       "      <td>161</td>\n",
       "      <td>0.827327</td>\n",
       "      <td>P08913</td>\n",
       "      <td>P78527</td>\n",
       "      <td>T161</td>\n",
       "      <td>IEYNLKRTPRRIKAI</td>\n",
       "      <td>P08913_161</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>42370 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Uniprot-Acc  blastp-identity kinase  position  combined-p-value  \\\n",
       "0          O43151            100.0   CDK1       396          0.001721   \n",
       "1          O43151            100.0   CDK1      1549          0.003608   \n",
       "2          O43151            100.0   CDK1       753          0.006664   \n",
       "3          O43151            100.0   CDK1      1000          0.007765   \n",
       "4          O43151            100.0   CDK1       428          0.008986   \n",
       "...           ...              ...    ...       ...               ...   \n",
       "42365      P08913            100.0  PRKDC       242          0.798014   \n",
       "42366      P08913            100.0  PRKDC       247          0.808342   \n",
       "42367      P08913            100.0  PRKDC       135          0.817593   \n",
       "42368      P08913            100.0  PRKDC       143          0.826284   \n",
       "42369      P08913            100.0  PRKDC       161          0.827327   \n",
       "\n",
       "      substrate_acc kinase_acc   site              pep substrate_id  \n",
       "0            O43151     P06493   S396  SPPAPFRSPQSYLRA   O43151_396  \n",
       "1            O43151     P06493  S1549  FNSALKGSPGFQDKL  O43151_1549  \n",
       "2            O43151     P06493   S753  ESPFATRSPKQIKIE   O43151_753  \n",
       "3            O43151     P06493  T1000  CKYARSKTPRKFRLA  O43151_1000  \n",
       "4            O43151     P06493   T428  SSAFPPATPRTEFPE   O43151_428  \n",
       "...             ...        ...    ...              ...          ...  \n",
       "42365        P08913     P78527   T242  YQIAKRRTRVPPSRR   P08913_242  \n",
       "42366        P08913     P78527   S247  RRTRVPPSRRGPDAV   P08913_247  \n",
       "42367        P08913     P78527   S135  DVLFCTSSIVHLCAI   P08913_135  \n",
       "42368        P08913     P78527   S143  IVHLCAISLDRYWSI   P08913_143  \n",
       "42369        P08913     P78527   T161  IEYNLKRTPRRIKAI   P08913_161  \n",
       "\n",
       "[42370 rows x 10 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_mapUpdateSite = pd.read_csv(PP_temp_dir_site_update + 'PICK_updated-9_mappedSite.csv')  \n",
    "df_mapUpdateSite = df_mapUpdateSite[(df_mapUpdateSite['site'].str.contains('S|T|Y', na=False)) & (df_mapUpdateSite['substrate_id'] != 'outdated')]\n",
    "df_mapUpdateSite.to_csv(PP_temp_dir_site + 'PICK_updated-9_mappedSite.csv')\n",
    "\n",
    "df_mapUpdateSite"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get the Gene Name of the Substrates from the Reference Human Proteome\n",
    "- get and add the Gene Name that would use across all perdictors for the substrates to the result files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>UniprotID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>ACTN1</td>\n",
       "      <td>P12814</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>STAT3</td>\n",
       "      <td>P40763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>ADD1</td>\n",
       "      <td>P35611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>ADD2</td>\n",
       "      <td>P35612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>ADRA2A</td>\n",
       "      <td>P08913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>20358</td>\n",
       "      <td>HSFX4</td>\n",
       "      <td>A0A1B0GTS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>20359</td>\n",
       "      <td>TRBJ2-6</td>\n",
       "      <td>A0A0A0MT70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>20360</td>\n",
       "      <td>TMEM225B</td>\n",
       "      <td>P0DP42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>20361</td>\n",
       "      <td>SMIM29</td>\n",
       "      <td>Q86T20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>20362</td>\n",
       "      <td>SCYGR6</td>\n",
       "      <td>A0A286YF77</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20363 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Gene Name   UniprotID\n",
       "0         ACTN1      P12814\n",
       "1         STAT3      P40763\n",
       "2          ADD1      P35611\n",
       "3          ADD2      P35612\n",
       "4        ADRA2A      P08913\n",
       "...         ...         ...\n",
       "20358     HSFX4  A0A1B0GTS1\n",
       "20359   TRBJ2-6  A0A0A0MT70\n",
       "20360  TMEM225B      P0DP42\n",
       "20361    SMIM29      Q86T20\n",
       "20362    SCYGR6  A0A286YF77\n",
       "\n",
       "[20363 rows x 2 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get the new list of unique kinase with common kinase name that would use across all referece and the uniprotID for these kinase\n",
    "df_unique_sub =  pd.read_csv(HP_csv, usecols = ['UniprotID','Gene Name'], sep = '\\t')\n",
    "df_unique_sub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uniprot-Acc</th>\n",
       "      <th>blastp-identity</th>\n",
       "      <th>kinase</th>\n",
       "      <th>position</th>\n",
       "      <th>combined-p-value</th>\n",
       "      <th>substrate_acc</th>\n",
       "      <th>kinase_acc</th>\n",
       "      <th>site</th>\n",
       "      <th>pep</th>\n",
       "      <th>substrate_id</th>\n",
       "      <th>Gene Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>P18077</td>\n",
       "      <td>100.0</td>\n",
       "      <td>CDK1</td>\n",
       "      <td>59</td>\n",
       "      <td>0.017008</td>\n",
       "      <td>P18077</td>\n",
       "      <td>P06493</td>\n",
       "      <td>T59</td>\n",
       "      <td>KAKNNTVTPGGKPNK</td>\n",
       "      <td>P18077_59</td>\n",
       "      <td>RPL35A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>P18077</td>\n",
       "      <td>100.0</td>\n",
       "      <td>CDK1</td>\n",
       "      <td>75</td>\n",
       "      <td>0.303455</td>\n",
       "      <td>P18077</td>\n",
       "      <td>P06493</td>\n",
       "      <td>T75</td>\n",
       "      <td>RVIWGKVTRAHGNSG</td>\n",
       "      <td>P18077_75</td>\n",
       "      <td>RPL35A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>P18077</td>\n",
       "      <td>100.0</td>\n",
       "      <td>CDK1</td>\n",
       "      <td>67</td>\n",
       "      <td>0.313442</td>\n",
       "      <td>P18077</td>\n",
       "      <td>P06493</td>\n",
       "      <td>T67</td>\n",
       "      <td>PGGKPNKTRVIWGKV</td>\n",
       "      <td>P18077_67</td>\n",
       "      <td>RPL35A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>P18077</td>\n",
       "      <td>100.0</td>\n",
       "      <td>CDK1</td>\n",
       "      <td>57</td>\n",
       "      <td>0.591686</td>\n",
       "      <td>P18077</td>\n",
       "      <td>P06493</td>\n",
       "      <td>T57</td>\n",
       "      <td>VYKAKNNTVTPGGKP</td>\n",
       "      <td>P18077_57</td>\n",
       "      <td>RPL35A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>P18077</td>\n",
       "      <td>100.0</td>\n",
       "      <td>CDK1</td>\n",
       "      <td>108</td>\n",
       "      <td>0.621989</td>\n",
       "      <td>P18077</td>\n",
       "      <td>P06493</td>\n",
       "      <td>S108</td>\n",
       "      <td>IRVMLYPSRI_____</td>\n",
       "      <td>P18077_108</td>\n",
       "      <td>RPL35A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6625186</td>\n",
       "      <td>Q9H2W6</td>\n",
       "      <td>100.0</td>\n",
       "      <td>PRKDC</td>\n",
       "      <td>214</td>\n",
       "      <td>0.819006</td>\n",
       "      <td>Q9H2W6</td>\n",
       "      <td>P78527</td>\n",
       "      <td>T214</td>\n",
       "      <td>NAPCGHYTFKFPQAM</td>\n",
       "      <td>Q9H2W6_214</td>\n",
       "      <td>MRPL46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6625187</td>\n",
       "      <td>Q9H2W6</td>\n",
       "      <td>100.0</td>\n",
       "      <td>PRKDC</td>\n",
       "      <td>196</td>\n",
       "      <td>0.876844</td>\n",
       "      <td>Q9H2W6</td>\n",
       "      <td>P78527</td>\n",
       "      <td>S196</td>\n",
       "      <td>ERTLATLSENNMEAK</td>\n",
       "      <td>Q9H2W6_196</td>\n",
       "      <td>MRPL46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6625188</td>\n",
       "      <td>Q9H2W6</td>\n",
       "      <td>100.0</td>\n",
       "      <td>PRKDC</td>\n",
       "      <td>147</td>\n",
       "      <td>0.899907</td>\n",
       "      <td>Q9H2W6</td>\n",
       "      <td>P78527</td>\n",
       "      <td>T147</td>\n",
       "      <td>ADEKNDRTSLNRKLD</td>\n",
       "      <td>Q9H2W6_147</td>\n",
       "      <td>MRPL46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6625189</td>\n",
       "      <td>Q9H2W6</td>\n",
       "      <td>100.0</td>\n",
       "      <td>PRKDC</td>\n",
       "      <td>148</td>\n",
       "      <td>0.918730</td>\n",
       "      <td>Q9H2W6</td>\n",
       "      <td>P78527</td>\n",
       "      <td>S148</td>\n",
       "      <td>DEKNDRTSLNRKLDR</td>\n",
       "      <td>Q9H2W6_148</td>\n",
       "      <td>MRPL46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6625190</td>\n",
       "      <td>Q9H2W6</td>\n",
       "      <td>100.0</td>\n",
       "      <td>PRKDC</td>\n",
       "      <td>138</td>\n",
       "      <td>0.925166</td>\n",
       "      <td>Q9H2W6</td>\n",
       "      <td>P78527</td>\n",
       "      <td>T138</td>\n",
       "      <td>FKLGARITEADEKND</td>\n",
       "      <td>Q9H2W6_138</td>\n",
       "      <td>MRPL46</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6625191 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Uniprot-Acc  blastp-identity kinase  position  combined-p-value  \\\n",
       "0            P18077            100.0   CDK1        59          0.017008   \n",
       "1            P18077            100.0   CDK1        75          0.303455   \n",
       "2            P18077            100.0   CDK1        67          0.313442   \n",
       "3            P18077            100.0   CDK1        57          0.591686   \n",
       "4            P18077            100.0   CDK1       108          0.621989   \n",
       "...             ...              ...    ...       ...               ...   \n",
       "6625186      Q9H2W6            100.0  PRKDC       214          0.819006   \n",
       "6625187      Q9H2W6            100.0  PRKDC       196          0.876844   \n",
       "6625188      Q9H2W6            100.0  PRKDC       147          0.899907   \n",
       "6625189      Q9H2W6            100.0  PRKDC       148          0.918730   \n",
       "6625190      Q9H2W6            100.0  PRKDC       138          0.925166   \n",
       "\n",
       "        substrate_acc kinase_acc  site              pep substrate_id Gene Name  \n",
       "0              P18077     P06493   T59  KAKNNTVTPGGKPNK    P18077_59    RPL35A  \n",
       "1              P18077     P06493   T75  RVIWGKVTRAHGNSG    P18077_75    RPL35A  \n",
       "2              P18077     P06493   T67  PGGKPNKTRVIWGKV    P18077_67    RPL35A  \n",
       "3              P18077     P06493   T57  VYKAKNNTVTPGGKP    P18077_57    RPL35A  \n",
       "4              P18077     P06493  S108  IRVMLYPSRI_____   P18077_108    RPL35A  \n",
       "...               ...        ...   ...              ...          ...       ...  \n",
       "6625186        Q9H2W6     P78527  T214  NAPCGHYTFKFPQAM   Q9H2W6_214    MRPL46  \n",
       "6625187        Q9H2W6     P78527  S196  ERTLATLSENNMEAK   Q9H2W6_196    MRPL46  \n",
       "6625188        Q9H2W6     P78527  T147  ADEKNDRTSLNRKLD   Q9H2W6_147    MRPL46  \n",
       "6625189        Q9H2W6     P78527  S148  DEKNDRTSLNRKLDR   Q9H2W6_148    MRPL46  \n",
       "6625190        Q9H2W6     P78527  T138  FKLGARITEADEKND   Q9H2W6_138    MRPL46  \n",
       "\n",
       "[6625191 rows x 11 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "start = time.time()\n",
    "all_results = glob.glob(PP_temp_dir_site + '*.csv')\n",
    "\n",
    "for filename in all_results:\n",
    "    df = pd.read_csv(filename)\n",
    "    # merge df_subsMap with df_unique_sub to add the common substrate gene name to the df\n",
    "    df = df.merge(df_unique_sub, left_on=['substrate_acc'], right_on=['UniprotID'], how = 'left')\n",
    "\n",
    "    df = df.drop(columns = ['UniprotID'])\n",
    "    df.to_csv(filename, index=False)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Creating Resource Files\n",
    "**globalKinaseMap**\n",
    "- creat a new or add unique kinases from NetworKIN to the globel kinase resource file.\n",
    "- get and add the Kinase Name that would use across all perdictors for the kinases to the result files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     kinase         kinase_acc\n",
      "67  RPSK6A5  (no hit in human)\n",
      "94   MAP3KB  (no hit in human)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>kinase</th>\n",
       "      <th>kinase_acc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>CDK1</td>\n",
       "      <td>P06493</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>CDK2</td>\n",
       "      <td>P24941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>CDK3</td>\n",
       "      <td>Q00526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>CDK4</td>\n",
       "      <td>P11802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>CDK5</td>\n",
       "      <td>Q00535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>102</td>\n",
       "      <td>VRK1</td>\n",
       "      <td>Q99986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>103</td>\n",
       "      <td>ATM</td>\n",
       "      <td>Q13315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>104</td>\n",
       "      <td>ATR</td>\n",
       "      <td>Q13535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>105</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>P42345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>106</td>\n",
       "      <td>PRKDC</td>\n",
       "      <td>P78527</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>107 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    kinase kinase_acc\n",
       "0     CDK1     P06493\n",
       "1     CDK2     P24941\n",
       "2     CDK3     Q00526\n",
       "3     CDK4     P11802\n",
       "4     CDK5     Q00535\n",
       "..     ...        ...\n",
       "102   VRK1     Q99986\n",
       "103    ATM     Q13315\n",
       "104    ATR     Q13535\n",
       "105   MTOR     P42345\n",
       "106  PRKDC     P78527\n",
       "\n",
       "[107 rows x 2 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get the kinases that failed to retrieve uniprotID\n",
    "all_results = glob.glob(PP_temp_dir_site + '*.csv')\n",
    "\n",
    "df_unique_kin = pd.DataFrame()\n",
    "\n",
    "for filename in all_results:\n",
    "    df = pd.read_csv(filename, usecols = ['kinase_acc', 'kinase'])\n",
    "    df = df.drop_duplicates()\n",
    "    df_unique_kin = df_unique_kin.append(df, ignore_index=True)\n",
    "    \n",
    "df_unique_kin = df_unique_kin.drop_duplicates()\n",
    "print (df_unique_kin[df_unique_kin['kinase_acc'] == '(no hit in human)'])\n",
    "df_unique_kin"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Manually check/search for the ones that fails to retrieve the UniprotID (column 'kinase_acc') programmatically:\n",
    "- there are 2 such substrates (Jan. 2020):\n",
    "    - RPSK6A5: may refer to RPS6KA5?\n",
    "    - MAP3KB : didn't find any record in human \n",
    "- Create a dictionary for the ones that need to manually enter the kinase_acc\n",
    "    - key = 'kinase' (the provided uniprotID)\n",
    "    - value = 'kinase_acc (the mapped uniprotID)\n",
    "\n",
    "|kinase|Mapped ID|\n",
    "|---|---|\n",
    "|RPSK6A5|O75582|\n",
    "|MAP3KB| |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "id_dict = {'RPSK6A5':'O75582'}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- add the uniprotID of the above protein kinases in the '*_mappedSite.csv' files\n",
    "- get a list of unique kinases in PhosphoPICK"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>kinase_acc</th>\n",
       "      <th>kinase</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>P06493</td>\n",
       "      <td>CDK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>P24941</td>\n",
       "      <td>CDK2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>Q00526</td>\n",
       "      <td>CDK3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>P11802</td>\n",
       "      <td>CDK4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>Q00535</td>\n",
       "      <td>CDK5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>101</td>\n",
       "      <td>Q99986</td>\n",
       "      <td>VRK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>102</td>\n",
       "      <td>Q13315</td>\n",
       "      <td>ATM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>103</td>\n",
       "      <td>Q13535</td>\n",
       "      <td>ATR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>104</td>\n",
       "      <td>P42345</td>\n",
       "      <td>MTOR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>105</td>\n",
       "      <td>P78527</td>\n",
       "      <td>PRKDC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>106 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    kinase_acc kinase\n",
       "0       P06493   CDK1\n",
       "1       P24941   CDK2\n",
       "2       Q00526   CDK3\n",
       "3       P11802   CDK4\n",
       "4       Q00535   CDK5\n",
       "..         ...    ...\n",
       "101     Q99986   VRK1\n",
       "102     Q13315    ATM\n",
       "103     Q13535    ATR\n",
       "104     P42345   MTOR\n",
       "105     P78527  PRKDC\n",
       "\n",
       "[106 rows x 2 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_results = glob.glob(PP_temp_dir_site + '*.csv')\n",
    "\n",
    "df_unique_kin = pd.DataFrame()\n",
    "\n",
    "for filename in all_results:\n",
    "    df = pd.read_csv(filename)\n",
    "    for key in id_dict:\n",
    "        df.loc[df.kinase == key, [\"kinase_acc\"]] = id_dict[key]\n",
    "    # remove MAP3KB : didn't find any record in human\n",
    "    df = df[df['kinase_acc'] != '(no hit in human)']\n",
    "    df.to_csv(filename, index = False)\n",
    "    df = df[['kinase_acc', 'kinase']]\n",
    "    df = df.drop_duplicates()\n",
    "    df_unique_kin = df_unique_kin.append(df, ignore_index=True)\n",
    "    \n",
    "df_unique_kin = df_unique_kin.drop_duplicates()\n",
    "\n",
    "df_unique_kin"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- creat a new or add unique kinases from PhosphoPICK to the globel kinase resource file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Empty DataFrame\n",
      "Columns: []\n",
      "Index: []\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Kinase Name</th>\n",
       "      <th>UniprotID</th>\n",
       "      <th>description</th>\n",
       "      <th>HPRD_kinase_name</th>\n",
       "      <th>HPRD_kinase_uniprot_id</th>\n",
       "      <th>HPRD_kinase_refseq_id</th>\n",
       "      <th>PhosphoSite_kinase_uniprot_id</th>\n",
       "      <th>PhosphoSite_kinase_name</th>\n",
       "      <th>PhosphoSite_kinase_gene_name</th>\n",
       "      <th>GPS5_kinase_name</th>\n",
       "      <th>NetworKIN_kinase_name</th>\n",
       "      <th>PhosphoPICK_kinase_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>SGK1</td>\n",
       "      <td>O00141</td>\n",
       "      <td>serum/glucocorticoid regulated kinase 1</td>\n",
       "      <td>SGK1</td>\n",
       "      <td>O00141</td>\n",
       "      <td>NP_001137148.1</td>\n",
       "      <td>O00141</td>\n",
       "      <td>SGK1</td>\n",
       "      <td>SGK1</td>\n",
       "      <td>SGK1</td>\n",
       "      <td>SGK1</td>\n",
       "      <td>SGK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>BMPR1B</td>\n",
       "      <td>O00238</td>\n",
       "      <td>bone morphogenetic protein receptor type 1B</td>\n",
       "      <td>BMPR1B</td>\n",
       "      <td>O00238</td>\n",
       "      <td>NP_001194.1</td>\n",
       "      <td>O00238</td>\n",
       "      <td>BMPR1B</td>\n",
       "      <td>BMPR1B</td>\n",
       "      <td>BMPR1B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>CDC7</td>\n",
       "      <td>O00311</td>\n",
       "      <td>cell division cycle 7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>O00311</td>\n",
       "      <td>CDC7</td>\n",
       "      <td>CDC7</td>\n",
       "      <td>CDC7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>PLK4</td>\n",
       "      <td>O00444</td>\n",
       "      <td>polo like kinase 4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>O00444</td>\n",
       "      <td>PLK4</td>\n",
       "      <td>PLK4</td>\n",
       "      <td>PLK4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>STK25</td>\n",
       "      <td>O00506</td>\n",
       "      <td>serine/threonine kinase 25</td>\n",
       "      <td>STK25</td>\n",
       "      <td>O00506</td>\n",
       "      <td>NP_006365.2</td>\n",
       "      <td>O00506</td>\n",
       "      <td>YSK1</td>\n",
       "      <td>STK25</td>\n",
       "      <td>STK25</td>\n",
       "      <td>STK25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>484</td>\n",
       "      <td>TP53RK</td>\n",
       "      <td>Q96S44</td>\n",
       "      <td>EKC/KEOPS complex subunit TP53RK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Q96S44</td>\n",
       "      <td>PRPK</td>\n",
       "      <td>TP53RK</td>\n",
       "      <td>TP53RK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>485</td>\n",
       "      <td>TRPM6</td>\n",
       "      <td>Q9BX84</td>\n",
       "      <td>Transient receptor potential cation channel su...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Q9BX84</td>\n",
       "      <td>ChaK2</td>\n",
       "      <td>TRPM6</td>\n",
       "      <td>TRPM6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>486</td>\n",
       "      <td>BCR/ABL</td>\n",
       "      <td>A9UF07</td>\n",
       "      <td>BCR/ABL fusion protein isoform Y5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A9UF07</td>\n",
       "      <td>BCR-ABL1</td>\n",
       "      <td>BCR/ABL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>487</td>\n",
       "      <td>BCKDK</td>\n",
       "      <td>O14874</td>\n",
       "      <td>[3-methyl-2-oxobutanoate dehydrogenase [lipoam...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BCKDK</td>\n",
       "      <td>BCKDK</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>488</td>\n",
       "      <td>EPHB6</td>\n",
       "      <td>O15197</td>\n",
       "      <td>Ephrin type-B receptor 6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EPHB6</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>489 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Kinase Name UniprotID                                        description  \\\n",
       "0          SGK1    O00141           serum/glucocorticoid regulated kinase 1    \n",
       "1        BMPR1B    O00238       bone morphogenetic protein receptor type 1B    \n",
       "2          CDC7    O00311                             cell division cycle 7    \n",
       "3          PLK4    O00444                                polo like kinase 4    \n",
       "4         STK25    O00506                        serine/threonine kinase 25    \n",
       "..          ...       ...                                                ...   \n",
       "484      TP53RK    Q96S44                   EKC/KEOPS complex subunit TP53RK   \n",
       "485       TRPM6    Q9BX84  Transient receptor potential cation channel su...   \n",
       "486     BCR/ABL    A9UF07                  BCR/ABL fusion protein isoform Y5   \n",
       "487       BCKDK    O14874  [3-methyl-2-oxobutanoate dehydrogenase [lipoam...   \n",
       "488       EPHB6    O15197                           Ephrin type-B receptor 6   \n",
       "\n",
       "    HPRD_kinase_name HPRD_kinase_uniprot_id HPRD_kinase_refseq_id  \\\n",
       "0               SGK1                 O00141        NP_001137148.1   \n",
       "1             BMPR1B                 O00238           NP_001194.1   \n",
       "2                NaN                    NaN                   NaN   \n",
       "3                NaN                    NaN                   NaN   \n",
       "4              STK25                 O00506           NP_006365.2   \n",
       "..               ...                    ...                   ...   \n",
       "484              NaN                    NaN                   NaN   \n",
       "485              NaN                    NaN                   NaN   \n",
       "486              NaN                    NaN                   NaN   \n",
       "487              NaN                    NaN                   NaN   \n",
       "488              NaN                    NaN                   NaN   \n",
       "\n",
       "    PhosphoSite_kinase_uniprot_id PhosphoSite_kinase_name  \\\n",
       "0                          O00141                    SGK1   \n",
       "1                          O00238                  BMPR1B   \n",
       "2                          O00311                    CDC7   \n",
       "3                          O00444                    PLK4   \n",
       "4                          O00506                    YSK1   \n",
       "..                            ...                     ...   \n",
       "484                        Q96S44                    PRPK   \n",
       "485                        Q9BX84                   ChaK2   \n",
       "486                        A9UF07                BCR-ABL1   \n",
       "487                           NaN                     NaN   \n",
       "488                           NaN                     NaN   \n",
       "\n",
       "    PhosphoSite_kinase_gene_name GPS5_kinase_name NetworKIN_kinase_name  \\\n",
       "0                           SGK1             SGK1                  SGK1   \n",
       "1                         BMPR1B           BMPR1B                   NaN   \n",
       "2                           CDC7             CDC7                   NaN   \n",
       "3                           PLK4             PLK4                   NaN   \n",
       "4                          STK25            STK25                 STK25   \n",
       "..                           ...              ...                   ...   \n",
       "484                       TP53RK           TP53RK                   NaN   \n",
       "485                        TRPM6            TRPM6                   NaN   \n",
       "486                      BCR/ABL              NaN                   NaN   \n",
       "487                          NaN            BCKDK                 BCKDK   \n",
       "488                          NaN              NaN                 EPHB6   \n",
       "\n",
       "    PhosphoPICK_kinase_name  \n",
       "0                      SGK1  \n",
       "1                       NaN  \n",
       "2                       NaN  \n",
       "3                       NaN  \n",
       "4                       NaN  \n",
       "..                      ...  \n",
       "484                     NaN  \n",
       "485                     NaN  \n",
       "486                     NaN  \n",
       "487                     NaN  \n",
       "488                     NaN  \n",
       "\n",
       "[489 rows x 12 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "start = time.time()\n",
    "unmapped_list = pd.DataFrame()\n",
    "\n",
    "# create df for the globalkinaseMap.csv if exist\n",
    "if os.path.isfile(KinaseMap): \n",
    "    df_humanKinase = pd.read_csv(KinaseMap)\n",
    "# if globalkinaseMap.csv file does not exist, create an new df using orginal human kinase map\n",
    "else:\n",
    "    df_humanKinase = pd.read_csv(HK_org, usecols = ['Kinase Name', 'Preferred Name', 'UniprotID', 'Type', 'description'], sep = '\\t')\n",
    "    df_humanKinase['description'].replace(regex=True,inplace=True,to_replace=r'\\[Source.+\\]',value=r'')\n",
    "\n",
    "for index, row in df_unique_kin.iterrows():\n",
    "    kinase = df_unique_kin.at[index, 'kinase_acc']\n",
    "    # if the kinase/other enzyme already in the globalKinaseMap.csv file\n",
    "    if any(df_humanKinase.UniprotID == kinase):\n",
    "        # get the index of the substrate in the globalKinaseMap.csv file \n",
    "        idx = df_humanKinase.index[df_humanKinase.UniprotID == kinase].values[0] \n",
    "\n",
    "        df_humanKinase.at[idx, 'PhosphoPICK_kinase_name'] = df_unique_kin.at[index, 'kinase']\n",
    "        \n",
    "    # if the kinase is not in the globalSubstrateMap.csv file, we need a list to check annotations manullay\n",
    "    else:\n",
    "        unmapped_list = unmapped_list.append(row,sort=False).reset_index(drop=True)\n",
    "        \n",
    "print (unmapped_list)\n",
    "df_humanKinase.to_csv(KinaseMap, index = False)\n",
    "df_humanKinase "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Kinase Name</th>\n",
       "      <th>UniprotID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>SGK1</td>\n",
       "      <td>O00141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>BMPR1B</td>\n",
       "      <td>O00238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>CDC7</td>\n",
       "      <td>O00311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>PLK4</td>\n",
       "      <td>O00444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>STK25</td>\n",
       "      <td>O00506</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>484</td>\n",
       "      <td>TP53RK</td>\n",
       "      <td>Q96S44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>485</td>\n",
       "      <td>TRPM6</td>\n",
       "      <td>Q9BX84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>486</td>\n",
       "      <td>BCR/ABL</td>\n",
       "      <td>A9UF07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>487</td>\n",
       "      <td>BCKDK</td>\n",
       "      <td>O14874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>488</td>\n",
       "      <td>EPHB6</td>\n",
       "      <td>O15197</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>489 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Kinase Name UniprotID\n",
       "0          SGK1    O00141\n",
       "1        BMPR1B    O00238\n",
       "2          CDC7    O00311\n",
       "3          PLK4    O00444\n",
       "4         STK25    O00506\n",
       "..          ...       ...\n",
       "484      TP53RK    Q96S44\n",
       "485       TRPM6    Q9BX84\n",
       "486     BCR/ABL    A9UF07\n",
       "487       BCKDK    O14874\n",
       "488       EPHB6    O15197\n",
       "\n",
       "[489 rows x 2 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get the new list kinase with common kinase name that would use across all referece and the uniprotID for these kinase\n",
    "df_unique_kin = df_humanKinase[['Kinase Name','UniprotID']]\n",
    "df_unique_kin"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- add the Kinase Name that would use across all perdictors for the kinase to the result files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uniprot-Acc</th>\n",
       "      <th>blastp-identity</th>\n",
       "      <th>kinase</th>\n",
       "      <th>position</th>\n",
       "      <th>combined-p-value</th>\n",
       "      <th>substrate_acc</th>\n",
       "      <th>kinase_acc</th>\n",
       "      <th>site</th>\n",
       "      <th>pep</th>\n",
       "      <th>substrate_id</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Kinase Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>P18077</td>\n",
       "      <td>100.0</td>\n",
       "      <td>CDK1</td>\n",
       "      <td>59</td>\n",
       "      <td>0.017008</td>\n",
       "      <td>P18077</td>\n",
       "      <td>P06493</td>\n",
       "      <td>T59</td>\n",
       "      <td>KAKNNTVTPGGKPNK</td>\n",
       "      <td>P18077_59</td>\n",
       "      <td>RPL35A</td>\n",
       "      <td>CDK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>P18077</td>\n",
       "      <td>100.0</td>\n",
       "      <td>CDK1</td>\n",
       "      <td>75</td>\n",
       "      <td>0.303455</td>\n",
       "      <td>P18077</td>\n",
       "      <td>P06493</td>\n",
       "      <td>T75</td>\n",
       "      <td>RVIWGKVTRAHGNSG</td>\n",
       "      <td>P18077_75</td>\n",
       "      <td>RPL35A</td>\n",
       "      <td>CDK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>P18077</td>\n",
       "      <td>100.0</td>\n",
       "      <td>CDK1</td>\n",
       "      <td>67</td>\n",
       "      <td>0.313442</td>\n",
       "      <td>P18077</td>\n",
       "      <td>P06493</td>\n",
       "      <td>T67</td>\n",
       "      <td>PGGKPNKTRVIWGKV</td>\n",
       "      <td>P18077_67</td>\n",
       "      <td>RPL35A</td>\n",
       "      <td>CDK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>P18077</td>\n",
       "      <td>100.0</td>\n",
       "      <td>CDK1</td>\n",
       "      <td>57</td>\n",
       "      <td>0.591686</td>\n",
       "      <td>P18077</td>\n",
       "      <td>P06493</td>\n",
       "      <td>T57</td>\n",
       "      <td>VYKAKNNTVTPGGKP</td>\n",
       "      <td>P18077_57</td>\n",
       "      <td>RPL35A</td>\n",
       "      <td>CDK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>P18077</td>\n",
       "      <td>100.0</td>\n",
       "      <td>CDK1</td>\n",
       "      <td>108</td>\n",
       "      <td>0.621989</td>\n",
       "      <td>P18077</td>\n",
       "      <td>P06493</td>\n",
       "      <td>S108</td>\n",
       "      <td>IRVMLYPSRI_____</td>\n",
       "      <td>P18077_108</td>\n",
       "      <td>RPL35A</td>\n",
       "      <td>CDK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6553695</td>\n",
       "      <td>Q9H2W6</td>\n",
       "      <td>100.0</td>\n",
       "      <td>PRKDC</td>\n",
       "      <td>214</td>\n",
       "      <td>0.819006</td>\n",
       "      <td>Q9H2W6</td>\n",
       "      <td>P78527</td>\n",
       "      <td>T214</td>\n",
       "      <td>NAPCGHYTFKFPQAM</td>\n",
       "      <td>Q9H2W6_214</td>\n",
       "      <td>MRPL46</td>\n",
       "      <td>PRKDC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6553696</td>\n",
       "      <td>Q9H2W6</td>\n",
       "      <td>100.0</td>\n",
       "      <td>PRKDC</td>\n",
       "      <td>196</td>\n",
       "      <td>0.876844</td>\n",
       "      <td>Q9H2W6</td>\n",
       "      <td>P78527</td>\n",
       "      <td>S196</td>\n",
       "      <td>ERTLATLSENNMEAK</td>\n",
       "      <td>Q9H2W6_196</td>\n",
       "      <td>MRPL46</td>\n",
       "      <td>PRKDC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6553697</td>\n",
       "      <td>Q9H2W6</td>\n",
       "      <td>100.0</td>\n",
       "      <td>PRKDC</td>\n",
       "      <td>147</td>\n",
       "      <td>0.899907</td>\n",
       "      <td>Q9H2W6</td>\n",
       "      <td>P78527</td>\n",
       "      <td>T147</td>\n",
       "      <td>ADEKNDRTSLNRKLD</td>\n",
       "      <td>Q9H2W6_147</td>\n",
       "      <td>MRPL46</td>\n",
       "      <td>PRKDC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6553698</td>\n",
       "      <td>Q9H2W6</td>\n",
       "      <td>100.0</td>\n",
       "      <td>PRKDC</td>\n",
       "      <td>148</td>\n",
       "      <td>0.918730</td>\n",
       "      <td>Q9H2W6</td>\n",
       "      <td>P78527</td>\n",
       "      <td>S148</td>\n",
       "      <td>DEKNDRTSLNRKLDR</td>\n",
       "      <td>Q9H2W6_148</td>\n",
       "      <td>MRPL46</td>\n",
       "      <td>PRKDC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6553699</td>\n",
       "      <td>Q9H2W6</td>\n",
       "      <td>100.0</td>\n",
       "      <td>PRKDC</td>\n",
       "      <td>138</td>\n",
       "      <td>0.925166</td>\n",
       "      <td>Q9H2W6</td>\n",
       "      <td>P78527</td>\n",
       "      <td>T138</td>\n",
       "      <td>FKLGARITEADEKND</td>\n",
       "      <td>Q9H2W6_138</td>\n",
       "      <td>MRPL46</td>\n",
       "      <td>PRKDC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6553700 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Uniprot-Acc  blastp-identity kinase  position  combined-p-value  \\\n",
       "0            P18077            100.0   CDK1        59          0.017008   \n",
       "1            P18077            100.0   CDK1        75          0.303455   \n",
       "2            P18077            100.0   CDK1        67          0.313442   \n",
       "3            P18077            100.0   CDK1        57          0.591686   \n",
       "4            P18077            100.0   CDK1       108          0.621989   \n",
       "...             ...              ...    ...       ...               ...   \n",
       "6553695      Q9H2W6            100.0  PRKDC       214          0.819006   \n",
       "6553696      Q9H2W6            100.0  PRKDC       196          0.876844   \n",
       "6553697      Q9H2W6            100.0  PRKDC       147          0.899907   \n",
       "6553698      Q9H2W6            100.0  PRKDC       148          0.918730   \n",
       "6553699      Q9H2W6            100.0  PRKDC       138          0.925166   \n",
       "\n",
       "        substrate_acc kinase_acc  site              pep substrate_id  \\\n",
       "0              P18077     P06493   T59  KAKNNTVTPGGKPNK    P18077_59   \n",
       "1              P18077     P06493   T75  RVIWGKVTRAHGNSG    P18077_75   \n",
       "2              P18077     P06493   T67  PGGKPNKTRVIWGKV    P18077_67   \n",
       "3              P18077     P06493   T57  VYKAKNNTVTPGGKP    P18077_57   \n",
       "4              P18077     P06493  S108  IRVMLYPSRI_____   P18077_108   \n",
       "...               ...        ...   ...              ...          ...   \n",
       "6553695        Q9H2W6     P78527  T214  NAPCGHYTFKFPQAM   Q9H2W6_214   \n",
       "6553696        Q9H2W6     P78527  S196  ERTLATLSENNMEAK   Q9H2W6_196   \n",
       "6553697        Q9H2W6     P78527  T147  ADEKNDRTSLNRKLD   Q9H2W6_147   \n",
       "6553698        Q9H2W6     P78527  S148  DEKNDRTSLNRKLDR   Q9H2W6_148   \n",
       "6553699        Q9H2W6     P78527  T138  FKLGARITEADEKND   Q9H2W6_138   \n",
       "\n",
       "        Gene Name Kinase Name  \n",
       "0          RPL35A        CDK1  \n",
       "1          RPL35A        CDK1  \n",
       "2          RPL35A        CDK1  \n",
       "3          RPL35A        CDK1  \n",
       "4          RPL35A        CDK1  \n",
       "...           ...         ...  \n",
       "6553695    MRPL46       PRKDC  \n",
       "6553696    MRPL46       PRKDC  \n",
       "6553697    MRPL46       PRKDC  \n",
       "6553698    MRPL46       PRKDC  \n",
       "6553699    MRPL46       PRKDC  \n",
       "\n",
       "[6553700 rows x 12 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "start = time.time()\n",
    "all_results = glob.glob(PP_temp_dir_site + '*.csv')\n",
    "\n",
    "for filename in all_results:\n",
    "    df = pd.read_csv(filename)\n",
    "    # merge with df_unique_kinase to add the common kinase name to the df\n",
    "    df = df.merge(df_unique_kin, left_on='kinase_acc', right_on='UniprotID', how = 'left')\n",
    "    # drop the duplicated uniprot id for kinases\n",
    "    df = df.drop(columns = 'UniprotID')\n",
    "\n",
    "    df.to_csv(filename,index=False)  \n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Standard Formatted PhosphoPICK\n",
    "### 'PICK_formatted.csv'\n",
    "Standardize the preprocessed file with following columns:\n",
    "- **substrate_id** - unique IDs for the substrate phosphorylation site (substrate_acc + position)\n",
    "- **substrate** - gene name for the substrates\n",
    "- **substrate_acc** - mapped UniprotIDs for the substrates\n",
    "- **site** - phosphorylation  site\n",
    "- **pep** - +/- 7 AA peptide sequence around the site\n",
    "- **score** - perdiction score\n",
    "- **kinase** - Kinase name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_results = glob.glob(PP_temp_dir_site + '*.csv')\n",
    "PP = []\n",
    "for filename in all_results:\n",
    "    df = pd.read_csv(filename, usecols = ['substrate_id', 'Gene Name','Uniprot-Acc','substrate_acc','site','pep', 'combined-p-value','Kinase Name'])\n",
    "    df = df.rename(columns={'Gene Name': 'substrate_name', 'combined-p-value': 'score'})\n",
    "    PP.append(df)\n",
    "df_final = pd.concat(PP)\n",
    "df_final = df_final.drop_duplicates()\n",
    "df_final = df_final.reset_index(drop=True)\n",
    "\n",
    "df_final"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- check for substrate/kinase prediction with multiple scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "duplicateRowsDF = df_final[df_final.duplicated(['substrate_id', 'Kinase Name'], keep = False)].sort_values(by=['substrate_id', 'Kinase Name'])\n",
    "duplicateRowsDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- get the index of the ones that the \"Uniprot-Acc\" (PhosphoPICK predicted substrate uniprotID) is a secondary UniprotID\n",
    "- remove them from the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sub_2nd_acc = duplicateRowsDF.index[duplicateRowsDF['Uniprot-Acc'] != duplicateRowsDF['substrate_acc']].tolist()\n",
    "sub_2nd_acc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_final = df_final.drop(df_final.index[sub_2nd_acc])\n",
    "df_final = df_final.reset_index(drop=True)\n",
    "df_final.to_csv(PP_formatted, chunksize = 1000000, index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
